backend/layers/processing/dataset_metadata_update.py

```diff
@@ -0,0 +1,284 @@
+"""
+Creates a new DatasetVersion to update metadata across dataset artifacts
+"""
+import json
+import logging
+import multiprocessing
+import os
+
+import scanpy
+import tiledb
+from rpy2.robjects import StrVector
+from rpy2.robjects.packages import importr
+
+from backend.common.utils.corpora_constants import CorporaConstants
+from backend.layers.business.business import BusinessLogic
+from backend.layers.common.entities import (
+    DatasetArtifactMetadataUpdate,
+    DatasetArtifactType,
+    DatasetConversionStatus,
+    DatasetProcessingStatus,
+    DatasetStatusKey,
+    DatasetValidationStatus,
+    DatasetVersion,
+    DatasetVersionId,
+)
+from backend.layers.persistence.persistence import DatabaseProvider
+from backend.layers.processing.h5ad_data_file import H5ADDataFile
+from backend.layers.processing.logger import configure_logging
+from backend.layers.processing.process_download import ProcessDownload
+from backend.layers.thirdparty.s3_provider import S3Provider
+from backend.layers.thirdparty.uri_provider import UriProvider
+
+configure_logging(level=logging.INFO)
+
+base = importr("base")
+seurat = importr("SeuratObject")
+
+# maps artifact name for metadata field to DB field name, if different
+ARTIFACT_TO_DB_FIELD = {"title": "name"}
+
+
+class DatasetMetadataUpdater(ProcessDownload):
+    def __init__(
+        self, business_logic: BusinessLogic, artifact_bucket: str, cellxgene_bucket: str, datasets_bucket: str
+    ) -> None:
+        super().__init__(business_logic, business_logic.uri_provider, business_logic.s3_provider)
+        self.artifact_bucket = artifact_bucket
+        self.cellxgene_bucket = cellxgene_bucket
+        self.datasets_bucket = datasets_bucket
+
+    def update_h5ad(
+        self,
+        h5ad_uri: str,
+        old_dataset_version: DatasetVersion,
+        new_key_prefix: str,
+        new_dataset_version_id: DatasetVersionId,
+        metadata_update: DatasetArtifactMetadataUpdate,
+    ):
+        h5ad_filename = self.download_from_source_uri(
+            source_uri=h5ad_uri,
+            local_path=CorporaConstants.LABELED_H5AD_ARTIFACT_FILENAME,
+        )
+
+        adata = scanpy.read_h5ad(h5ad_filename, backed="r")
+        metadata = old_dataset_version.metadata
+        # maps artifact name for metadata field to DB field name, if different
+        for key, val in metadata_update.as_dict_without_none_values().items():
+            adata.uns[key] = val
+
+            db_field = ARTIFACT_TO_DB_FIELD.get(key) if key in ARTIFACT_TO_DB_FIELD else key
+            setattr(metadata, db_field, val)
+
+        adata.write(h5ad_filename, compression="gzip")
+        self.business_logic.set_dataset_metadata(new_dataset_version_id, metadata)
+
+        self.create_artifact(
```

Comment:
After we upload the artifacts we should delete the files locally downloaded and generated. This is to avoid running out of disk space on the EC2.